gobblin-runtime/src/main/java/gobblin/runtime/Fork.java

```diff
@@ -90,17 +115,83 @@ public Fork(TaskContext taskContext, TaskState taskState, Object schema, int bra
     this.rowLevelPolicyCheckingResult = new RowLevelPolicyCheckResults();
     this.writer = buildWriter();
 
+    this.recordQueue = BoundedBlockingRecordQueue.newBuilder()
+        .hasCapacity(taskState.getPropAsInt(
+            ConfigurationKeys.FORK_BRANCH_RECORD_QUEUE_CAPACITY_KEY,
+            ConfigurationKeys.DEFAULT_FORK_BRANCH_RECORD_QUEUE_CAPACITY))
+        .useTimeout(taskState.getPropAsLong(
+            ConfigurationKeys.FORK_BRANCH_RECORD_QUEUE_TIMEOUT_KEY,
+            ConfigurationKeys.DEFAULT_FORK_BRANCH_RECORD_QUEUE_TIMEOUT))
+        .useTimeoutTimeUnit(TimeUnit.valueOf(taskState.getProp(
+            ConfigurationKeys.FORK_BRANCH_RECORD_QUEUE_TIMEOUT_UNIT_KEY,
+            ConfigurationKeys.DEFAULT_FORK_BRANCH_RECORD_QUEUE_TIMEOUT_UNIT)))
+        .collectStats()
+        .build();
+
+    this.countDownLatch = countDownLatch;
+
     this.closer.register(this.rowLevelPolicyChecker);
     this.closer.register(this.writer);
+
+    this.forkState = new AtomicReference<ForkState>(ForkState.PENDING);
   }
 
-  public void processRecord(Object record)
-      throws Exception {
-    for (Object convertedRecord : this.converter.convertRecord(this.convertedSchema, record, this.taskState)) {
-      if (this.rowLevelPolicyChecker.executePolicies(convertedRecord, this.rowLevelPolicyCheckingResult)) {
-        this.writer.write(convertedRecord);
-      }
+  @Override
+  public void run() {
+    setForkState(ForkState.PENDING, ForkState.RUNNING);
+    try {
+      processRecords();
+      setForkState(ForkState.RUNNING, ForkState.SUCCEEDED);
+    } catch (IOException ioe) {
+      setForkState(ForkState.RUNNING, ForkState.FAILED);
```

Comment:
I think setting state for FAILED explicitly is better. If setForkState() throws an exception, it will obfuscate the real exception -- ioe.